### PR TITLE
GPII-1222: Decompose template handling components to more clearly separate their responsibilities...

### PR DESCRIPTION
**Pull request by the-t-in-rtf**
In working with the `templateFormControl` component, I was finding it very hard to handle the simple use case of a component that needs to retrieve its data before rendering.  Rather than continue creating components that disable portions of the base component, I separated out the request handling portions from `templateFormControl` and `templateAware`, so that it is now possible to use the same rule-based engine to handle an initial "get my data" request and subsequent "save my changes" requests.  For more details, see [GPII-1122](https://issues.gpii.net/browse/GPII-1222).

As part of this work, the handlebars client template was updated so that you can either supply templates to it through its options, or retrieve them from the server.

### CODE REVIEW

**Comment by the-t-in-rtf** [comment]
@amb26, this pull request has not been touched yet, it's awaiting your feedback.

**Comment by amb26** [comment]
It looks like all of this content is included in https://github.com/GPII/gpii-handlebars/pull/10 where I have reviewed it

File: tests/js/zombie-binder-tests.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                var updatedFormElement = component.locate("updateFromModel");
+                var clientSideFluid = browser.window.fluid;
+                var clientSideValue = clientSideFluid.value(updatedFormElement);
+                jqUnit.assertEquals("A change to the model should result in an update to the form element", "updated using applier", clientSideValue);
+            });
+        });
+    });
+};
+
+var testFile = "file://" + path.resolve(__dirname, "../html/tests-binder.html");
```
Comment (amb26) [inline tests/js/zombie-binder-tests.js]: This file manipulation is not portable. These tests are failing for me on Windows with the following:

result  { module: 'Testing binder separately from template handling...',
  name: 'Testing component \'long\'...',
  result: false,
  message: 'There should be no errors:Error: Server returned status code 404 from file://e/source/gits/gpii-handlebars/tests/html/tests-binder.html',
  source: '    at pok (E:\source\gits\gpii-handlebars\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:102:15)\n    at Object.jsUnitCompat.fa
il (E:\source\gits\gpii-handlebars\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:119:13)\n    at Browser.<anonymous> (E:\source\gits\gp
ii-handlebars\tests\js\zombie-binder-tests.js:32:24)\n    at Browser.emit (events.js:117:20)\n    at E:\source\gits\gpii-handlebars\node_modules\zombie\
\lib\document.js:127:21\n    at E:\source\gits\gpii-handlebars\node_modules\zombie\lib\eventloop.js:305:11\n    at Object._onImmediate (E:\source\gits
\gpii-handlebars\node_modules\zombie\lib\eventloop.js:616:11)\n    at processImmediate [as _immediateCallback](timers.js:354:15)',
  test: 'Testing component \'long\'...' }
jq:  FAIL: Module "Testing binder separately from template handling..." Test name "Testing component 'long'..." - Message: There should be no errors:Error: Serv
er returned status code 404 from file://e/source/gits/gpii-handlebars/tests/html/tests-binder.html
jq:  Source:     at pok (E:\source\gits\gpii-handlebars\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:102:15)
    at Object.jsUnitCompat.fail (E:\source\gits\gpii-handlebars\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:119:13)
    at Browser.<anonymous> (E:\source\gits\gpii-handlebars\tests\js\zombie-binder-tests.js:32:24)
    at Browser.emit (events.js:117:20)
    at E:\source\gits\gpii-handlebars\node_modules\zombie\lib\document.js:127:21
    at E:\source\gits\gpii-handlebars\node_modules\zombie\lib\eventloop.js:305:11
    at Object._onImmediate (E:\source\gits\gpii-handlebars\node_modules\zombie\lib\eventloop.js:616:11)
    at processImmediate [as _immediateCallback](timers.js:354:15)

**Comment by amb26** [comment]
Thanks for all the work to fix up the tests on Windows - they all now pass for me. Mysteriously though, I get two EADDRINUSE diagnostics before the tests start to execute - here a transcript:

01:50:48.303:  Express server listening on port 6984
01:50:48.304:  Express started...
01:50:48.306:  FATAL ERROR: Uncaught exception: listen EADDRINUSE
Error: listen EADDRINUSE
    at errnoException (net.js:905:11)
    at Server._listen2 (net.js:1043:14)
    at listen (net.js:1065:10)
    at Server.listen (net.js:1139:5)
    at EventEmitter.app.listen (E:\source\gits\gpii-handlebars\node_modules\gpii-express\node_modules\express\lib\application.js:559:24)
    at gpii.express.init (E:\source\gits\gpii-handlebars\node_modules\gpii-express\src\js\express.js:118:32)
    at Object.fluid.event.invokeListener (E:\source\gits\gpii-handlebars\node_modules\infusion\src\framework\core\js\Fluid.js:1052:25)
    at togo (E:\source\gits\gpii-handlebars\node_modules\infusion\src\framework\core\js\FluidIoC.js:1713:36)
    at fireToListeners (E:\source\gits\gpii-handlebars\node_modules\infusion\src\framework\core\js\Fluid.js:1103:68)
    at that.fire (E:\source\gits\gpii-handlebars\node_modules\infusion\src\framework\core\js\Fluid.js:1196:24)
    at Object.fluid.fireEvent (E:\source\gits\gpii-handlebars\node_modules\infusion\src\framework\core\js\Fluid.js:1210:24)
    at Object.fluid.initComponent (E:\source\gits\gpii-handlebars\node_modules\infusion\src\framework\core\js\Fluid.js:2295:15)
    at Object.creator (E:\source\gits\gpii-handlebars\node_modules\infusion\src\framework\core\js\Fluid.js:1575:26)
    at Object.<anonymous> (E:\source\gits\gpii-handlebars\tests\js\zombie-binder-tests.js:70:29)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
01:50:48.339:  Express server listening on port 6994
01:50:48.340:  Express started...
01:50:48.344:  Express server listening on port 6995
01:50:48.345:  Express started...
01:50:48.347:  FATAL ERROR: Uncaught exception: listen EADDRINUSE
Error: listen EADDRINUSE
    at errnoException (net.js:905:11)
    at Server._listen2 (net.js:1043:14)
    at listen (net.js:1065:10)
    at Server.listen (net.js:1139:5)
    at EventEmitter.app.listen (E:\source\gits\gpii-handlebars\node_modules\gpii-express\node_modules\express\lib\application.js:559:24)
    at gpii.express.init (E:\source\gits\gpii-handlebars\node_modules\gpii-express\src\js\express.js:118:32)
    at Object.fluid.event.invokeListener (E:\source\gits\gpii-handlebars\node_modules\infusion\src\framework\core\js\Fluid.js:1052:25)
    at togo (E:\source\gits\gpii-handlebars\node_modules\infusion\src\framework\core\js\FluidIoC.js:1713:36)
    at fireToListeners (E:\source\gits\gpii-handlebars\node_modules\infusion\src\framework\core\js\Fluid.js:1103:68)
    at that.fire (E:\source\gits\gpii-handlebars\node_modules\infusion\src\framework\core\js\Fluid.js:1196:24)
    at Object.fluid.fireEvent (E:\source\gits\gpii-handlebars\node_modules\infusion\src\framework\core\js\Fluid.js:1210:24)
    at Object.fluid.initComponent (E:\source\gits\gpii-handlebars\node_modules\infusion\src\framework\core\js\Fluid.js:2295:15)
    at Object.creator (E:\source\gits\gpii-handlebars\node_modules\infusion\src\framework\core\js\Fluid.js:1575:26)
    at Object.<anonymous> (E:\source\gits\gpii-handlebars\tests\js\zombie-templateFormControl-tests.js:135:29)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
01:50:48.375:  Express server listening on port 6904

What's strange is that these uncaught exceptions don't seem to prevent the test sequences from running, and also, that I don't have any services running on ports in that range. Perhaps somehow the fixtures are trying to start up multiple instances of the same servers? In any case - could you add an exception handler within gpii-express that will dump which port the failed listen is on as well as other details of the failed activity. My sense is that two servers try to listen on port 6984 - and that the fault is that the all-tests.js driver is failing to serialise the execution of the multiple test fixtures properly. If you were to switch over to the IoC testing framework, this is something that it would take care of automatically - but there will no doubt be a way of achieving this by other means - perhaps by having a test fixture file returning a promise bracketing its start and finish operation and then feeding these into an algorithm like fluid.promise.sequence or when.sequence

**Comment by the-t-in-rtf** [comment]
Yes, when I said all tests passed, I meant if you run them individually.  We will need some kind of headless browser test component that plays nicely with the IoC testing framework, otherwise even our best pass would be some kind of a hybrid with the Zombie bits in nested callbacks or chained promises.  For now, I will do as you suggest and wrap the existing tests in promises.

**Comment by the-t-in-rtf** [comment]
@amb26, I figured out the problem.  A few tests used the same express port number.  Even with each script returning a promise, the previous instance is not completely destroyed before the next starts.  The tests were succeeding in your example because they were run against the previous express instance before it was destroyed.

To address this, I added a utility which wraps the resolve function from when.js in a configurable timeout (500 ms by default).  With this, all tests run in sequence and each test does not start until the previous express instance has been completely destroyed.

**Comment by the-t-in-rtf** [comment]
The harness now takes care of creating and resolving the promise as suggested.  All tests that use the harness have been updated.

**Pull request merged**